anchor-service/src/anchor.rs

```diff
@@ -76,15 +100,123 @@ pub struct TimeEventBatch {
     /// The anchor proof
     pub proof: Proof,
     /// The Time Events
-    pub time_events: TimeEvents,
+    pub raw_time_events: RawTimeEvents,
 }
 
 impl std::fmt::Debug for TimeEventBatch {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("TimeEventBatch")
             .field("merkle_nodes", &self.merkle_nodes)
             .field("proof", &self.proof)
-            .field("time_events", &self.time_events)
+            .field("raw_time_events", &self.raw_time_events)
             .finish()
     }
 }
+
+impl TimeEventBatch {
+    /// Try to convert the TimeEventBatch into a list of TimeEventInsertables
+    pub fn try_to_insertables(&self) -> Result<Vec<TimeEventInsertable>> {
+        info!(
+            "store anchor batch: proof={}, events={}",
+            self.proof.to_cid()?,
+            self.raw_time_events.events.len()
+        );
+        let events = self
+            .raw_time_events
+            .events
+            .iter()
+            .map(|(anchor_request, time_event)| {
+                self.build_time_event_insertable(time_event, anchor_request)
+            })
+            .collect::<Result<Vec<_>>>()?;
+        Ok(events)
+    }
+
+    /// Build a TimeEventInsertable from a RawTimeEvent and AnchorRequest
+    fn build_time_event_insertable(
+        &self,
+        time_event: &RawTimeEvent,
```

Comment:
should this be owned since we're cloning out of it? it seems like a move makes sense and the semantic of consuming raw and returning regular seems reasonable.

Reply:
We would have to work around the partial move.

```rs
error[E0382]: borrow of partially moved value: `self`
   --> anchor-service/src/anchor.rs:128:18
    |
127 |             .into_iter()
    |              ----------- `self.raw_time_events.events` partially moved due to this method call
128 |             .map(|(anchor_request, time_event)| {
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after partial move
129 |                 self.build_time_event_insertable(time_event, anchor_request)
    |                 ---- borrow occurs due to use in closure
    |
```

try_to_insertables calls self.build_time_event_insertable
build_time_event_insertable calls self.find_tree_blocks_along_path

Reply:
yeah, I mean we might have to still call clone, and it might not even change the generated code. I mostly meant whether the API semantic should be to take ownership and consume them into something else, but if it doesn't help the internal code it's fine. 

I just generally like API design that doesn't have internal clones and let's the higher level decide whether to move/clone, but sometimes when you only need a partial field it's better to borrow the whole thing and simply clone the smaller pieces out.